File: tests/fixtures/cover-behaviour-gate/reads_source.py
# Fixture for tests/unit/test-cover-behaviour-gate.mjs: rule (c), reading source instead of running it.
import inspect
import subprocess
from pathlib import Path

from app import pricing
from app.pricing import apply_discount

SOURCE = Path("app/pricing.py").read_text()


def test_defines_apply_discount():
    assert "def apply_discount" in SOURCE


def test_guard_is_present():
    assert "rate < 0" in inspect.getsource(pricing.apply_discount)


def test_no_print_calls():
    result = subprocess.run(["grep", "-n", "print(", "app/pricing.py"], capture_output=True)
    assert result.returncode == 1


def test_opens_the_module():
    with open("app/pricing.py") as fh:
        assert "Decimal" in fh.read()


def test_still_computes_a_total():
    assert apply_discount(10, 0.5) == 5

File: tests/fixtures/cover-behaviour-gate/reads-source.ts
// Fixture for tests/unit/test-cover-behaviour-gate.mjs: rule (c), reading source instead of running it.
import { execSync } from 'node:child_process';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { expect, it } from 'vitest';
import { applyDiscount } from '../src/pricing';

const PRICING = join(__dirname, '..', 'src', 'pricing.ts');
const SRC = readFileSync(PRICING, 'utf8');

it('exports applyDiscount', () => {
  expect(SRC).toContain('export function applyDiscount');
});

it('has no TODO left in pricing', () => {
  const hits = execSync('grep -c TODO src/pricing.ts || true').toString();
  expect(hits.trim()).toBe('0');
});

it('guards the rate inline', () => {
  const text = readFileSync('src/pricing.ts', 'utf8');
  expect(text).toMatch(/rate < 0/);
});

it('still computes a total', () => {
  expect(applyDiscount(10, 0.5)).toBe(5);
});
